### PR TITLE
#42572: check for program config that cannot be handled by matmul kernels

### LINKED ISSUE #42572: ttnn.matmul hangs with fuse_batch=False + batched weight + block-sharded output
## Summary

`ttnn.matmul` with `MatmulMultiCoreReuseMultiCastProgramConfig(fuse_batch=False)` **hangs** (no error, no timeout) when:

1. Both inputs have a non-unit batch dim (`in1` batch != 1), so `fuse_batch=False` is the only legal choice — `fuse_batch=True` is rejected by the shape check at `matmul_device_operation.cpp:393`, AND
2. The output `memory_config` is **block-sharded on the compute grid**.

Switching the output memory_config to interleaved L1 makes the same call complete in milliseconds.

Expected: the matmul either completes successfully or is rejected at `validate()` time with a descriptive error.

## Environment

- Hardware: Wormhole B0, single device
- tt-metal: `v0.69.0-dev20260414-8-g90c914ef25`
- Python 3.12, Ubuntu 24.04

## Minimal reproducer

```python
import torch
import ttnn

device = ttnn.open_device(device_id=0)
try:
    in0_t = torch.randn((4, 32, 2880), dtype=torch.bfloat16)
    in1_t = torch.randn((4, 2880, 5760), dtype=torch.bfloat16)

    in0 = ttnn.from_torch(
        in0_t, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT,
        device=device, memory_config=ttnn.L1_MEMORY_CONFIG,
    )
    in1 = ttnn.from_torch(
        in1_t, dtype=ttnn.bfloat8_b, layout=ttnn.TILE_LAYOUT,
        device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG,
    )

    program_config = ttnn.MatmulMultiCoreReuseMultiCastProgramConfig(
        compute_with_storage_grid_size=ttnn.CoreCoord(8, 4),
        in0_block_w=2,
        out_subblock_h=1, out_subblock_w=1,
        out_block_h=1,   out_block_w=23,
        per_core_M=1,    per_core_N=23,
        transpose_mcast=False,
        fuse_batch=False,
    )

    block_sharded_out = ttnn.MemoryConfig(
        memory_layout=ttnn.TensorMemoryLayout.BLOCK_SHARDED,
        buffer_type=ttnn.BufferType.L1,
        shard_spec=ttnn.ShardSpec(
            ttnn.CoreRangeSet(
                [ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(7, 3))]
            ),
            (1 * 32, 23 * 32),
            ttnn.ShardOrientation.ROW_MAJOR,
        ),
    )

    out = ttnn.matmul(
        in0, in1,
        program_config=program_config,
        dtype=ttnn.bfloat16,
        memory_config=block_sharded_out,    # replace with ttnn.L1_MEMORY_CONFIG -> completes
    )
    ttnn.synchronize_device(device)
    print("Matmul done.", out.shape)
finally:
    ttnn.close_device(device)
```

### PR DESCRIPTION
**Pull request by bbradelTT**
### Summary

Closes #42572

Forge team found a program config that resulted in a hang for matmul. The kernels were not designed for that combination of inputs, and there was no validation.

Added validation.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:bbradel-42572_mm_check)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:bbradel-42572_mm_check)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:bbradel-42572_mm_check)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:bbradel-42572_mm_check)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:bbradel-42572_mm_check)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:bbradel-42572_mm_check)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/matmul/device/factory/matmul_multicore_reuse_mcast_2d_program_factory.cpp [hunk 1]
```diff
@@ -106,6 +106,15 @@ MatmulMultiCoreReuseMcast2DProgramFactory::cached_program_t create_program_mcast
...
 
+    TT_FATAL(
+        !(output_is_sharded && B > 1),
+        "Block-sharded output is incompatible with batch > 1 (B={}). The output CB is backed by the shard buffer "
+        "which only holds per_core_M * per_core_N = {} tiles, but the kernel would produce B * per_core_M * per_core_N "
+        "= {} tiles without draining. Use fuse_batch=True.",
+        B,
+        per_core_M * per_core_N,
+        B * per_core_M * per_core_N);
+
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply ttnn/cpp/ttnn/operations/matmul/device/factory/matmul_multicore_reuse_mcast_2d_program_factory.cpp:117]: Keeping this one.

File: ttnn/cpp/ttnn/operations/matmul/device/factory/matmul_multicore_reuse_mcast_2d_program_factory.cpp [hunk 2]
```diff
@@ -1709,6 +1718,17 @@ static MatmulMultiCoreReuseMcast2DProgramFactory::cached_program_t matmul_multi_
...
+    if (B > 1 && out_buffer->buffer_layout() == tt_metal::TensorMemoryLayout::BLOCK_SHARDED) {
+        TT_FATAL(
+            false,
+            "Matmul with block-sharded output and fuse_batch=False is not supported when batch > 1 (batch={}). "
+            "Each core's output shard can only hold one batch worth of tiles, but the kernel would attempt to write "
+            "all {} batches into the same shard, causing a hang. Use fuse_batch=True to distribute batches across "
+            "the grid rows.",
+            B,
+            B);
+    }
```
[the remark this answers is not in the document]
Comment (bbradelTT) [reply ttnn/cpp/ttnn/operations/matmul/device/factory/matmul_multicore_reuse_mcast_2d_program_factory.cpp]: Removing this one.

**Review by edwinleeTT — APPROVED** [review]

**Review by fplavecTT — APPROVED** [review]
